lang/java/avro/src/main/java/org/apache/avro/generic/GenericDatumReader.java

```diff
@@ -234,8 +234,16 @@ protected Object convert(Object datum, Schema schema, LogicalType type, Conversi
    * representations.
    */
   protected Object readRecord(Object old, Schema expected, ResolvingDecoder in) throws IOException {
-    final Object record = data.newRecord(old, expected);
-    final Object state = data.getRecordState(record, expected);
+    final Schema realSchema;
+    if (expected.hasChild()) {
+      int index = in.readExtends();
+      realSchema = expected.findInHierachy(index);
+    } else {
+      realSchema = expected;
+    }
```

Comment:
This seems to be at odds with GenericDatumWriter, lines 239-241: there, the extends is optional, based on whether the parent schema or one of its children is written.

Reply:
indeed, i have to add unit test and fix that.

Reply:
You were right, so now writer will always write the "extends" symbol if schema has child.